File: experiments/adrianco/experiment-24-qwennext80b-cached/bookshop/runs/agent=hermes-local_language=typescript_prompt=neutral/rep2/src/controllers/BookController.ts
import { Request, Response } from 'express';
import { bookRepository } from '../database/BookRepository';
import { Book } from '../models/Book';

export class BookController {
  static getAll(req: Request, res: Response): void {
    const { author } = req.query;
    
    if (author && typeof author === 'string') {
      const books = bookRepository.getByAuthor(author);
      res.json(books);
      return;
    }

    const books = bookRepository.getAll();
    res.json(books);
  }

  static getById(req: Request, res: Response): void {
    const { id } = req.params;
    const book = bookRepository.getById(parseInt(id, 10));

    if (!book) {
      res.status(404).json({ error: 'Book not found' });
      return;
    }

    res.json(book);
  }

  static create(req: Request, res: Response): void {
    const book = bookRepository.create(req.body);
    res.status(201).json(book);
  }

  static update(req: Request, res: Response): void {
    const { id } = req.params;
    const book = bookRepository.update(parseInt(id, 10), req.body);

    if (!book) {
      res.status(404).json({ error: 'Book not found' });
      return;
    }

    res.json(book);
  }

  static delete(req: Request, res: Response): void {
    const { id } = req.params;
    const deleted = bookRepository.delete(parseInt(id, 10));

    if (!deleted) {
      res.status(404).json({ error: 'Book not found' });
      return;
    }

    res.status(204).send();
  }
}

File: experiments/adrianco/experiment-41-repair-80b-fullctx/bookshop/runs/agent=hermes-local_language=java_model=mlxlocal/mlx-community--Qwen3-Coder-Next-4bit_prompt=repair_stack=m80/rep3/src/main/java/com/bookapi/controller/BookController.java
package com.bookapi.controller;

import com.bookapi.model.Book;
import com.bookapi.repository.BookRepository;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Optional;

import static spark.Spark.*;

public class BookController {
    private static final Logger logger = LoggerFactory.getLogger(BookController.class);
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private final BookRepository bookRepository;

    public BookController() {
        this.bookRepository = new BookRepository();
        setupRoutes();
    }

    private void setupRoutes() {
        // Health check
        get("/health", (req, res) -> {
            res.status(200);
            return objectMapper.writeValueAsString("{\"status\": \"healthy\"}");
        });

        // List all books
        get("/books", (req, res) -> {
            String author = req.queryParams("author");
            List<Book> books = bookRepository.findAll(author);
            res.status(200);
            return objectMapper.writeValueAsString(books);
        });

        // Get single book
        get("/books/:id", (req, res) -> {
            int id = Integer.parseInt(req.params("id"));
            Optional<Book> book = bookRepository.findById(id);
            if (book.isPresent()) {
                res.status(200);
                return objectMapper.writeValueAsString(book.get());
            } else {
                res.status(404);
                return objectMapper.writeValueAsString("{\"error\": \"Book not found\"}");
            }
        });

        // Create book
        post("/books", (req, res) -> {
            try {
                Book book = objectMapper.readValue(req.body(), Book.class);
                Book savedBook = bookRepository.save(book);
                if (savedBook != null) {
                    res.status(201);
                    res.header("Location", "/books/" + savedBook.getId());
                    return objectMapper.writeValueAsString(savedBook);
                } else {
                    res.status(500);
                    return objectMapper.writeValueAsString("{\"error\": \"Failed to create book\"}");
                }
            } catch (Exception e) {
                logger.error("Error creating book", e);
                res.status(400);
                return objectMapper.writeValueAsString("{\"error\": \"Invalid request body\"}");
            }
        });

        // Update book
        put("/books/:id", (req, res) -> {
            try {
                int id = Integer.parseInt(req.params("id"));
                Optional<Book> existingBook = bookRepository.findById(id);
                if (!existingBook.isPresent()) {
                    res.status(404);
                    return objectMapper.writeValueAsString("{\"error\": \"Book not found\"}");
                }
                
                Book book = objectMapper.readValue(req.body(), Book.class);
                book.setId(id);
                Book updatedBook = bookRepository.save(book);
                
                res.status(200);
                return objectMapper.writeValueAsString(updatedBook);
            } catch (Exception e) {
                logger.error("Error updating book", e);
                res.status(400);
                return objectMapper.writeValueAsString("{\"error\": \"Invalid request body\"}");
            }
        });

        // Delete book
        delete("/books/:id", (req, res) -> {
            int id = Integer.parseInt(req.params("id"));
            boolean deleted = bookRepository.delete(id);
            if (deleted) {
                res.status(204);
            } else {
                res.status(404);
                return objectMapper.writeValueAsString("{\"error\": \"Book not found\"}");
            }
            return "";
        });
    }
}
